File: secure.py
import os
from flask import Flask, request, jsonify, render_template_string
import sqlite3  # Example DB
from markupsafe import escape

app = Flask(__name__)

# ✅ Disable debug mode in production
app.config['DEBUG'] = False

# ✅ Load DB credentials from environment variables
DB_PATH = os.getenv("DB_PATH", "test.db")  # Use environment variable or default

# Helper function to get DB connection
def get_db_connection():
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

@app.route('/user/<int:user_id>')
def get_user(user_id):
    """
    ✅ Use parameterized queries to prevent SQL injection
    """
    conn = get_db_connection()
    user = conn.execute("SELECT username, email FROM users WHERE id = ?", (user_id,)).fetchone()
    conn.close()

    if user:
        return jsonify({"username": user["username"], "email": user["email"]})
    else:
        return jsonify({"error": "User not found"}), 404

@app.route('/ping')
def ping():
    """
    ✅ Escape output to prevent XSS
    """
    message = request.args.get("message", "pong")
    safe_message = escape(message)
    return render_template_string("<p>{{ message }}</p>", message=safe_message)

# ✅ No os.system or subprocess calls -> prevent command injection

if __name__ == '__main__':
    # ✅ Production-safe: debug=False
    app.run(host='0.0.0.0', port=5000, debug=False)

File: secure.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class SecureApp {

    public static void main(String[] args) {
        // ✅ Load credentials from environment variables or secure vaultt
        String dbUser = System.getenv("DB_USER");
        String dbPass = System.getenv("DB_PASS");
        String dbUrl  = System.getenv("DB_URL"); // e.g., jdbc:mysql://localhost:3306/testdb

        if (dbUser == null || dbPass == null || dbUrl == null) {
            System.err.println("Database credentials missing! Set DB_USER, DB_PASS, DB_URL.");
            return;
        }

        try (Connection conn = DriverManager.getConnection(dbUrl, dbUser, dbPass)) {

            // ✅ Use parameterized queries to prevent SQL injection
            String query = "SELECT username, email FROM users WHERE id = ?";
            try (PreparedStatement stmt = conn.prepareStatement(query)) {
                int userId = 1; // Example safe input (never concatenate user input!)
                stmt.setInt(1, userId);

                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        System.out.println("Username: " + rs.getString("username"));
                        System.out.println("Email: " + rs.getString("email"));
                    }
                }
            }

            // ✅ Avoid any hardcoded secrets or passwords in code
            // Secrets should always be retrieved from environment variables, vaults, or encrypted storage

            // ✅ No insecure deserialization, no command execution, no debug mode

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
